assertion error in typer with 0.1.2-RC1

On [this file](https://github.com/rkuhn/akka-typed-session/blob/9727b432947cb7f69b95b0ad4e9265516068f5d1/src/main/scala/com/rolandkuhn/akka_typed_session/FSM.scala) dotc bails out with the assertion error below. Commenting out lines 135–142 changes the behavior to an assertion error in ScalaDSL.scala.

~~~
> launchIDE
[info] Compiling 10 Scala sources to /Users/rkuhn/comp/akka-typed-session/target/scala-0.1/classes...
[info] 
[info] exception occurred while typechecking /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/FSM.scala
[info] 
[info] exception occurred while compiling /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/Actor.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/Effects.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/FSM.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/MapAdapter.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/Operation.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/Process.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/ScalaDSL.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/State.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/internal/ProcessImpl.scala, /Users/rkuhn/comp/akka-typed-session/src/main/scala/com/rolandkuhn/akka_typed_session/package.scala
[trace] Stack trace suppressed: run last compile:compileIncremental for the full output.
[error] (compile:compileIncremental) java.lang.AssertionError: assertion failed
[error] Total time: 1 s, completed 2017-jun-03 20:51:00
~~~

The full stack trace:

~~~
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:234)
	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:534)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:615)
	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:671)
	at dotty.tools.dotc.typer.Applications.op$64$$anonfun$2(Applications.scala:696)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:1691)
	at dotty.tools.dotc.typer.Applications.op$64(Applications.scala:707)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:641)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:741)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1568)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:924)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1081)
	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1145)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:733)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:777)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:755)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:221)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:199)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:338)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1521)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1546)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1613)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1651)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1671)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:614)
	at dotty.tools.dotc.typer.Typer.op$99(Typer.scala:617)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:616)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1575)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.op$109(Typer.scala:805)
	at dotty.tools.dotc.typer.Typer.typedFunction(Typer.scala:687)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1577)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$975(ProtoTypes.scala:237)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:215)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:237)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:616)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:616)
	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:399)
	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:444)
	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:238)
	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:534)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:615)
	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:671)
	at dotty.tools.dotc.typer.Applications.op$64$$anonfun$2(Applications.scala:696)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:1691)
	at dotty.tools.dotc.typer.Applications.op$64(Applications.scala:707)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:641)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:741)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1568)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Typer.typeSelectOnTerm$1(Typer.scala:393)
	at dotty.tools.dotc.typer.Typer.op$83(Typer.scala:427)
	at dotty.tools.dotc.typer.Typer.typedSelect(Typer.scala:391)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1549)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1613)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Applications.op$64(Applications.scala:643)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:641)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:741)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1568)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Typer.typeSelectOnTerm$1(Typer.scala:393)
	at dotty.tools.dotc.typer.Typer.op$83(Typer.scala:427)
	at dotty.tools.dotc.typer.Typer.typedSelect(Typer.scala:391)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1549)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1613)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Applications.op$64(Applications.scala:643)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:641)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:741)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1568)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Typer.typeSelectOnTerm$1(Typer.scala:393)
	at dotty.tools.dotc.typer.Typer.op$83(Typer.scala:427)
	at dotty.tools.dotc.typer.Typer.typedSelect(Typer.scala:391)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1549)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1613)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Typer.op$99(Typer.scala:624)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:616)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1575)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:924)
	at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:934)
	at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1081)
	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1145)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:733)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:777)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:755)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:221)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:199)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:338)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1521)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1546)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1613)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1651)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1671)
	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:614)
	at dotty.tools.dotc.typer.Typer.op$99(Typer.scala:617)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:616)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1575)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.op$109(Typer.scala:805)
	at dotty.tools.dotc.typer.Typer.typedFunction(Typer.scala:687)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1577)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.Typer.op$99(Typer.scala:624)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:616)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1575)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$975(ProtoTypes.scala:237)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:215)
	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:237)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:616)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:616)
	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:399)
	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:444)
	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:238)
	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:534)
	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:615)
	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:671)
	at dotty.tools.dotc.typer.Applications.op$64$$anonfun$2(Applications.scala:696)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:1691)
	at dotty.tools.dotc.typer.Applications.op$64(Applications.scala:707)
	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:641)
	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:741)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1568)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1662)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1671)
	at dotty.tools.dotc.typer.Typer.op$106(Typer.scala:1357)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1283)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1559)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1613)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.traverse$4(Typer.scala:1651)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1671)
	at dotty.tools.dotc.typer.Typer.op$86(Typer.scala:1474)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1461)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1598)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1614)
	at dotty.tools.dotc.typer.Typer.op$116(Typer.scala:1629)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1683)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck$$anonfun$1(FrontEnd.scala:64)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:32)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:68)
	at dotty.tools.dotc.typer.FrontEnd.runOn$$anonfun$3(FrontEnd.scala:93)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:93)
	at dotty.tools.dotc.Run.$anonfun$$anonfun$10(Run.scala:82)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:90)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:95)
	at dotty.tools.dotc.Run.compileSources(Run.scala:64)
	at dotty.tools.dotc.Run.compile(Run.scala:48)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at xsbt.CachedCompilerImpl.run(CompilerInterface.scala:63)
	at xsbt.CachedCompilerImpl.run(CompilerInterface.scala:53)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:37)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:107)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:53)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
	at sbt.compiler.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:49)
	at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
	at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:32)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:72)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:71)
	at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:99)
	at sbt.inc.Incremental$.compile(Incremental.scala:71)
	at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
	at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
	at sbt.Compiler$.compile(Compiler.scala:155)
	at sbt.Compiler$.compile(Compiler.scala:141)
	at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:886)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:877)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:875)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
~~~

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -217,7 +217,7 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
         // apply the result type constraint, unless method type is dependent
         val resultApprox = resultTypeApprox(methType)
         val savedConstraint = ctx.typerState.constraint
-        if (!constrainResult(resultApprox, resultType) 
+        if (!constrainResult(resultApprox, resultType))
           if (ctx.typerState.isCommittable)
             // defer the problem until after the application;
             // it might be healed by an implicit conversion

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -216,7 +216,6 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
       case methType: MethodType =>
         // apply the result type constraint, unless method type is dependent
         val resultApprox = resultTypeApprox(methType)
-        val savedConstraint = ctx.typerState.constraint
         if (!constrainResult(resultApprox, resultType))
           if (ctx.typerState.isCommittable)
             // defer the problem until after the application;

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -217,11 +217,11 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
         // apply the result type constraint, unless method type is dependent
         val resultApprox = resultTypeApprox(methType)
         val savedConstraint = ctx.typerState.constraint
-        if (!constrainResult(resultApprox, resultType))
+        if (!constrainResult(resultApprox, resultType) 
           if (ctx.typerState.isCommittable)
             // defer the problem until after the application;
             // it might be healed by an implicit conversion
-            assert(ctx.typerState.constraint eq savedConstraint)
+            ()
           else
             fail(err.typeMismatchMsg(methType.resultType, resultType))
         // match all arguments with corresponding formal parameters

--- a/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
@@ -54,20 +54,23 @@ object ProtoTypes {
     /** Check that the result type of the current method
      *  fits the given expected result type.
      */
-    def constrainResult(mt: Type, pt: Type)(implicit ctx: Context): Boolean = pt match {
-      case pt: FunProto =>
-        mt match {
-          case mt: MethodType =>
-            constrainResult(resultTypeApprox(mt), pt.resultType)
-          case _ =>
-            true
-        }
-      case _: ValueTypeOrProto if !disregardProto(pt) =>
-        isCompatible(normalize(mt, pt), pt)
-      case pt: WildcardType if pt.optBounds.exists =>
-        isCompatible(normalize(mt, pt), pt)
-      case _ =>
-        true
+    def constrainResult(mt: Type, pt: Type)(implicit ctx: Context): Boolean = {
+      val savedConstraint = ctx.typerState.constraint
+      val res = pt match {
+        case pt: FunProto =>
+          mt match {
+            case mt: MethodType => constrainResult(resultTypeApprox(mt), pt.resultType)
+            case _ => true
+          }
+        case _: ValueTypeOrProto if !disregardProto(pt) =>
+          isCompatible(normalize(mt, pt), pt)
+        case pt: WildcardType if pt.optBounds.exists =>
+          isCompatible(normalize(mt, pt), pt)
+        case _ =>
+          true
+      }
+      if (!res) ctx.typerState.constraint = savedConstraint
+      res
     }
   }
 

--- a/tests/neg/i2672.scala
+++ b/tests/neg/i2672.scala
@@ -1,0 +1,16 @@
+class Foo[T]
+
+trait Prepend {
+  type Out
+}
+
+object Test {
+  def foo()(implicit ev: Prepend): Foo[ev.Out] = ???
+
+  def test: Unit = {
+    foo(): Foo[Any] // error: found: Prepend => Foo[_] required: Foo[Any]
+
+  }
+
+  implicit val p: Prepend = ???
+}

--- a/tests/pos/i2672.scala
+++ b/tests/pos/i2672.scala
@@ -1,0 +1,15 @@
+class Foo[+T]
+
+trait Prepend {
+  type Out
+}
+
+object Test {
+  def foo()(implicit ev: Prepend): Foo[ev.Out] = ???
+
+  def test: Unit = {
+    foo(): Foo[Any]
+  }
+
+  implicit val p: Prepend = ???
+}

